store/model.go

```diff
@@ -0,0 +1,276 @@
+	Before interface{}
+	After  interface{}
+}
+
+func (t *Txn) Create(new interface{}) error {
+	if t.readonly {
+		return ErrReadonlyTx
+	}
+	id := getEntityID(new)
+	if id == es.EmptyEntityID {
+		id = setNewEntityID(new)
+	}
+	key := t.model.dsKey.ChildString(id.String())
+	exists, err := t.model.datastore.Has(key)
+	if err != nil {
+		return err
+	}
+	if exists {
+		return errCantCreateExistingInstance
+	}
+
+	a := es.Action{
+		Type:       es.Create,
+		EntityID:   id,
+		EntityType: t.model.schema.Ref,
+		Previous:   nil,
+		Current:    new,
+	}
+	t.actions = append(t.actions, a)
+
+	return nil
+}
+
+func (t *Txn) Save(updated interface{}) error {
+	if t.readonly {
+		return ErrReadonlyTx
+	}
+
+	id := getEntityID(updated)
+	key := t.model.dsKey.ChildString(id.String())
+	beforeBytes, err := t.model.datastore.Get(key)
+	if err == ds.ErrNotFound {
+		return errCantSaveNonExistentInstance
+	}
+	if err != nil {
+		return err
+	}
+
+	before := reflect.New(t.model.valueType.Elem()).Interface()
+	err = json.Unmarshal(beforeBytes, before)
+	if err != nil {
+		return err
+	}
+	a := es.Action{
+		Type:       es.Save,
+		EntityID:   id,
+		EntityType: t.model.schema.Ref,
+		Previous:   before,
+		Current:    updated,
+	}
+	t.actions = append(t.actions, a)
+
+	return nil
+}
+
+func (t *Txn) Delete(id es.EntityID) error {
+	if t.readonly {
+		return ErrReadonlyTx
+	}
+	key := t.model.dsKey.ChildString(id.String())
+	exists, err := t.model.datastore.Has(key)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		return ErrNotFound
+	}
+	a := es.Action{
+		Type:       es.Delete,
+		EntityID:   id,
+		EntityType: t.model.schema.Ref,
+		Previous:   nil,
+		Current:    nil,
+	}
+	t.actions = append(t.actions, a)
+	return nil
+}
+
+func (t *Txn) Has(id es.EntityID) (bool, error) {
+	key := t.model.dsKey.ChildString(id.String())
+	exists, err := t.model.datastore.Has(key)
+	if err != nil {
+		return false, err
+	}
+	return exists, nil
+}
+
+func (t *Txn) FindByID(id es.EntityID, v interface{}) error {
+	key := t.model.dsKey.ChildString(id.String())
+	bytes, err := t.model.datastore.Get(key)
+	if errors.Is(err, ds.ErrNotFound) {
+		return ErrNotFound
+	}
+	if err != nil {
+		return err
+	}
+	return json.Unmarshal(bytes, v)
+}
+
+func (t *Txn) Commit() error {
+	if t.discarded || t.commited {
+		return errAlreadyDiscardedCommitedTxn
+	}
+
+	events, err := t.model.eventcreator.Create(t.actions)
+	if err != nil {
+		return err
+	}
+
+	for _, e := range events {
+		if err := t.model.dispatcher.Dispatch(e); err != nil {
+			return err // ToDo/Note: Important to document the implications of a partial dispatch
+		}
+	}
+	return nil
+}
+
+func (m *Model) Reduce(event es.Event) error {
+	// ToDo: distinguish local and remote events for proper locking
+	log.Debugf("reducer %s start", m.schema.Ref)
+	if event.Type() != m.schema.Ref {
+		log.Debugf("ignoring event from uninteresting type")
+		return nil
+	}
+
+	return m.eventcreator.Reduce(event, m.datastore, m.dsKey)
```

Comment:
This is something that feels a bit weird, since the `Model.Reduce` is mostly delegating the work to the `EventCreator`. Since a priori the `Model` can be used with a JSON-patcher, or JSON-CRDT or who knows which other implementation... so the `Model` doesn't really know what was encoded in the `Event` payload.

I went back and forth with making the `EventCreator` the real `Reducer` interface implementation... but I had some conflicting feelings since that would mean that the `EventCreator` would be the owner of the `datastore`... 

Not fully unconvinced, but if it makes sense to anyone to have some opinion, it would be appreciated. :sweat_smile:

Reply:
Mapping components to logical concern makes sense to me even if that means some are small, though I don't have an intuition here.

Reply:
You've convinced _me_ that having the EventCreator be the Reducer... seems like the right abstraction to me from the outside?